mmdet3d/core/bbox/coders/monoflex_bbox_coder.py

```diff
@@ -0,0 +1,387 @@
+        pred_depth_uncertainty = \
+            reg[:, 49:50].squeeze(-1)
+        # predictions for keypoints
+        pred_keypoints2d = reg[:, 6:26]
+        # (B * max_objs, 3)
+        pred_keypoints_depth = self.decode_depth_from_keypoints(
+            pred_keypoints2d, pred_dimensions, cam2imgs, down_ratio)
+        # predict the uncertainties of the solved depths from
+        # groups of keypoints
+        pred_corner_offset_uncertainty = \
+            reg[:, 26:29]
+        if self.use_combined_depth:
+            pred_combined_uncertainty = torch.cat(
+                (pred_depth_uncertainty.unsqueeze(-1),
+                 pred_corner_offset_uncertainty),
+                dim=1).exp()
+            pred_combined_depth = torch.cat(
+                (pred_direct_depth.unsqueeze(-1), pred_keypoints_depth), dim=1)
+            pred_uncertainty_weights = 1 / pred_combined_uncertainty
+            pred_uncertainty_weights = \
+                pred_uncertainty_weights / \
+                pred_uncertainty_weights.sum(dim=1, keepdim=True)
+            pred_corner_depth = torch.sum(
+                pred_combined_depth * pred_uncertainty_weights, dim=1)
+
+        # compute the corners
+        pred_locations = self.decode_location(points, pred_offsets2d,
+                                              pred_corner_depth, cam2imgs,
+                                              down_ratio)
+        # decode yaws and alphas
+        pred_orientations, _ = self.decode_orientation(pred_orientations,
+                                                       pred_locations)
+
+        return pred_locations, pred_dimensions, pred_orientations,
+
+    def decode_direct_depth(self, depth_offsets):
+        """Transform depth offset to depth.
+
+        Args:
+            depth_offsets (torch.Tensor): Predicted depth offsets.
+                shape: (B * max_objs, )
+
+        Return:
+            torch.Tensor: Directly regressed depth.
+                shape: (B * max_objs, )
+        """
+        if self.depth_mode == 'exp':
+            depth = depth_offsets.exp()
+        elif self.depth_mode == 'linear':
+            base_depth = depth_offsets.new_tensor(self.base_depth)
+            depth = depth_offsets * base_depth[1] + base_depth[0]
+        elif self.depth_mode == 'inv_sigmoid':
+            depth = 1 / torch.sigmoid(depth_offsets) - 1
+        else:
+            raise ValueError
+
+        if self.depth_range is not None:
+            depth = torch.clamp(
+                depth, min=self.depth_range[0], max=self.depth_range[1])
+
+        return depth
+
+    def decode_location(self,
+                        centers2d_target,
+                        offsets2d,
+                        depths,
+                        cam2imgs,
+                        down_ratio,
+                        pad_mode='default'):
+        """Retrieve object location.
+
+        Args:
+            centers2d_target (torch.Tensor): Projected 3D target centers
```

Comment:
From my understanding, does it mean base_centers2d? We can get the predicted centers2d from base_centers2d and offsets2d? centers2d_target can be a little confusing.